### DISCUSSION TITLE
How is everyone else doing iOS debugging

### DISCUSSION THREAD

**Discussion by williambuchanan2** [dotnet/maui#21320] [General]
Just wondering what the options are for this now?

VS for Mac doesn't work.
VS Code + Dev Kit doesn't work.
VS Hot Reload doesn't work.

So, how is everyone doing it? At the moment we are having to deploy through Devops to TestFlight to see the result of our changes.

**Comment by JohnHDev** [comment]
VS for Mac works, but I normally use VS on Windows paired to a Mac.

In what way does it not work for you?

**Reply by williambuchanan2** [reply]
Interesting that you say VS for Mac works, but maybe I should be more specific... The pattern with VS for mac has been that it works some days and then suddenly stops working (even if nothing has changed). Mostly (but not always) the problems are build failures - random errors that prevent the build from succeeding. However, after the latest .NET update we have been unable to run the apps (they build ok but just crash as soon as they load). All the above is happening consistently to the entire team so I would be interested if this isn't the case with others.

VS Windows paired with Mac (hot reload) - I managed to do this for a 'hello world' test app, but so far for both our real apps it has never worked. Mostly we get build errors, but often the problem is the app deploys with a white icon and then refuses to run because it cannot be trusted. Again, the whole team have burnt through a lot of hours on this and just gave up in the end.

VS Code + Dev Kit - the solution selector doesn't show when opening a folder with a solution in it, so just simply doesn't work. Also I just don't see why this is being put forward as a viable option for building apps. It is just a glorified text editor at best.

I have raised these problems a few times in bug reports but they don't seem to go anywhere, so literally no options at the moment.

**Comment by wearemeeters-admin** [comment]
Rider ;)

**Comment by JohnHDev** [comment]
I was going to suggest using Rider, but I suspect it will have the same issues. If so, it might be a problem with the Debug configuration, have you created new configurations or has anything been changed? I would start the new Hello World and see if that works, and move forward.
Rider has a free trial licence option, you could try it at least.
And just checking, you have done a fresh Clone from git to a clean folder? (Im presuming git or some other version control).

**Reply by williambuchanan2** [reply]
This is happening across all the team members, and I recently got a new macbook and the problem exists there as well.

But again - it's not a single problem. It is every day running into new problems - week after week lost due to it. And then it suddenly starts working again for a day or 2 (although at the moment it seems to have a terminal illness). With VS Code I have kind of got it working now, but it's a terrible experience and feels like I am back in the 1990's again.

Looking at your response it sounds like you aren't experiencing any of this...? This is the thing that surprises me.

**Comment by Ghostbird** [comment]
I'm using VSCode + .NET‌‌ Meteor plugin + iPhone simulator on a mac mini. It works fairly well. In some things it works better than Android debugging (with a similar set-up). The main problem is the limitations of the iPhone Simulator. I haven't managed to set-up physical device debugging yet. I don't understand all these hoops Apple wants me to jump through to get that working.

So yeah, we deploy to Testflight using fastlane (from that mac mini) to test on a physical device.

**Reply by williambuchanan2** [reply]
What seems to happen to us is we work on the simulator for a while, then it suddenly stops working - so we try physical device and that works. Then physical device debug stops working but we get simulator working again. We keep switching back and forth like that. I don't think we have ever had both working at the same time and you never know from one day to the next which one, if any, is going to work on that day.

**Reply by Ghostbird** [reply]
Is that a Simulator issue, or a MAUI issue? For us the Simulator has never stopped working, but it is missing some features, such as Camera and NFC. Though that did help prevent a bug from going to production recently. It only happened on devices that did not support NFC, and all of our physical test devices supported NFC, and were unaffected.

**Reply by williambuchanan2** [reply]
I am guessing it is a Maui/VS/.Net issue. The simulator runs ok, but the problems are always caused by random build or deploy errors, or runtime problems. It's different all the time.

**Comment by cleardemon** [comment]
I don't have an issue with iOS debugging, using MAUI 8.0.14 and debugging using iOS Simulator (17.2) and an actual iPhone 12 just fine.

My development machine is a MacBook M2 Pro, no Windows involved. I use Rider as my main IDE, but VS Code also works.

What sort of problems are you having?

**Reply by williambuchanan2** [reply]
It's hard to succinctly summarise what problems we are having. Every day it's a different problem. Mostly build or deploy failures. A lot of issues in VS for mac (stupid things like the selected debugger changing half way through building; build failures appearing despite nothing changing, etc). VS Code also seems to have a lot of issues. In general it's just a very difficult, unreliable, and time consuming process, and this is being experienced by the whole team.

**Comment by williambuchanan2** [comment]
I just found out that XCode 17.3 is not yet supported, so that explains my latest issue. It looks like that version of xcode and the latest version of Maui got released around the same time so I had assumed that both would be in sync. Turns out this isn't the case so we need to wait until the next Maui update to get it working again.

That said, I have got things working in VS Code on iPhone 14 at the moment so i'll just run with that until it stops working.

**Comment by williambuchanan2** [comment]
Just thought I would follow up on this. I finally took everyone's advice and tried Rider - would have saved myself a lot of pain if I had done it sooner! I kept hoping that MS would eventually fix the problems...

The whole team have had continued issues using VS Code or pair to mac - it is so unreliable and VS Code is a terrible tool to use.

Rider just works - you don't need to mess around with it.

**Reply by gabsamples6** [reply]
I would like to try rider is there a walkthrough of what you have to do to get rider working on Maui on Mac?

**Reply by Jay-57blocks** [reply]
@williambuchanan2 iOS would like to try rider is there a walkthrough of what you have to do to get rider working on Maui on Mac?

**Comment by phillippschmedt** [comment]
I am using Rider. But same issue. Debugging will SOMETIMES for no apparent reason not work anymore. But Android is worse. It happens to my collegues too. It seems be a problem that everyone experiences eventually but its just part of the .NET MAUI experience.

**Comment by LeoJHarris** [comment]
I'm experiencing excruciatingly long MAUI iOS build time atm, seriously, like I can barely do more then 4 redeployment in an hour on making any .csproj file or code changes. Maybe there's an element for me right now on just trying to find the sweet spot in the release config, `AOTCompilation` seriously doesn't work, I mean I haven't been at my desk long enough to see this even complete the build, I updated my MAUI workload this morning to 9.0.51 today and IDK if this has impacted something, but then again simple clean and rebuilds can sort of resolve project specific issues but sometimes it just feels likes its all a bit of trial and error until its actually working then not touching it again!

**Reply by DawidBester** [reply]
If you are building via Pair to Mac from a Windows machine, check the network latency between the Windows and Mac machines. I had a WiFi connection between the computers on my desk and didn't realize it introduced 66ms+ latency, which in turn resulted in very long iOS build times. 

The solution was to _setup a 2nd network as a point to point Ethernet connection between the Windows and Mac machines using private IP addresses_; the latency on the point to point network is 1ms. To use the point to point network for builds; from the Windows machine use Pair to Mac to connect to the Macs point to point private IP address, also forget the Mac WiFi IP address so its not longer used.  

iOS debug builds are now quick enough for me to be usable and the remote desktop display of iOS emulator running on Mac works consistently.  

Would be nice if Pair to Mac generated a high network latency warning as an alert. 
In my case the Pair to Mac icon always indicated green so I thought the network connection was good....
